File: src/input/core/socd.c
#include <assert.h>
#include <stddef.h>   /* NULL */
#include "socd.h"

/*
 * One SOCD axis (a pair of opposing directions). Returns the bits to KEEP for the
 * axis and updates *win with the current winner. A faithful port of the per-axis
 * loop in ../../input/socd.py -- keep the two in lockstep.
 */
static buttons_t socd_axis(socd_mode_t mode, buttons_t pressed, buttons_t prev,
                           buttons_t bit_a, buttons_t bit_b, buttons_t *win)
{
    int ca = (pressed & bit_a) != 0, cb = (pressed & bit_b) != 0;
    int pa = (prev & bit_a) != 0,    pb = (prev & bit_b) != 0;
    buttons_t keep = 0;

    if (ca && cb) {                                  /* SOCD conflict on this axis */
        switch (mode) {
        case SOCD_BYPASS:
            keep = bit_a | bit_b;
            break;
        case SOCD_NEUTRAL:
            keep = 0;                                /* neither survives            */
            break;
        case SOCD_UP_PRIORITY:
            keep = (bit_a == BTN_UP) ? bit_a : 0;    /* Up wins; horizontal -> 0    */
            break;
        case SOCD_LAST_WIN:
            if (ca && !pa && cb && !pb) *win = 0;     /* both same frame -> no winner (neutral) */
            else if (ca && !pa)         *win = bit_a; /* newest single press takes the axis */
            else if (cb && !pb)         *win = bit_b;
            keep = (*win == bit_a ? bit_a : 0) | (*win == bit_b ? bit_b : 0);
            break;
        case SOCD_FIRST_WIN:
            /* first press to stand alone wins and holds; a same-frame tie from neutral
             * has no "first" -> stays neutral (win==0) until one direction stands alone */
            keep = (*win == bit_a ? bit_a : 0) | (*win == bit_b ? bit_b : 0);
            break;
        }
    } else if (ca) {
        keep = bit_a; *win = bit_a;
    } else if (cb) {
        keep = bit_b; *win = bit_b;
    } else {
        *win = 0;                                    /* axis released -> no winner  */
    }
    return keep;
}

void socd_init(socd_t *s, socd_mode_t mode)
{
    assert(s != NULL);
    assert(mode <= SOCD_BYPASS);      /* enum is unsigned; SOCD_NEUTRAL==0 is the floor */
    s->mode = mode;
    s->prev = 0;
    s->win_v = 0;
    s->win_h = 0;
}

buttons_t socd_clean(socd_t *s, buttons_t pressed)
{
    assert(s != NULL);

    buttons_t keep_v = socd_axis(s->mode, pressed, s->prev, BTN_UP, BTN_DOWN, &s->win_v);
    buttons_t keep_h = socd_axis(s->mode, pressed, s->prev, BTN_LEFT, BTN_RIGHT, &s->win_h);

    /* non-direction buttons pass through untouched; only the four dirs are gated */
    buttons_t out = (pressed & ~BTN_DIRS) | keep_v | keep_h;
    s->prev = pressed;

    /* negative space: after cleaning, opposing directions never both survive
     * (bypass is the one mode that deliberately lets them through). */
    assert(s->mode == SOCD_BYPASS || (out & (BTN_UP | BTN_DOWN)) != (BTN_UP | BTN_DOWN));
    assert(s->mode == SOCD_BYPASS || (out & (BTN_LEFT | BTN_RIGHT)) != (BTN_LEFT | BTN_RIGHT));
    return out;
}

File: test/reference/input/socd.py
"""
SOCD cleaning — the input-side processor that resolves Simultaneous Opposite
Cardinal Directions (Left+Right, Up+Down) before they reach the shared buffer.

This is the *source* side of the buffer+reflector architecture (ARCHITECTURE.md):
raw button state -> SOCD clean -> buffer -> every reflector. Tournament rulesets
require SOCD cleaning on leverless controllers, and getting it deterministic +
zero-added-latency is the "first stick with hardware SOCD" claim
(docs/roadmap/CHIP-EXPLOITATION.md). This software cleaner is the reference logic;
on silicon it runs in the PIOC edge window / on the V3F core, off the hot path.

Modes:
  neutral       both opposing directions cancel to nothing
  up_priority   Up+Down -> Up (a common ruleset); Left+Right -> neutral
  last_win      the most recently pressed direction wins  (SECOND_INPUT_PRIORITY)
  first_win     the first pressed direction wins, held    (FIRST_INPUT_PRIORITY)
  bypass        no cleaning (raw dual-press passes through)

`last_win`/`first_win` are stateful (depend on press order), so the cleaner keeps
a tiny bit of per-axis state across calls. A same-frame tie from neutral (both
opposing directions arrive in one frame with no prior single press) has no
distinguishable order, so it resolves to neutral until one direction stands alone.
"""

MODES = ("neutral", "up_priority", "last_win", "first_win", "bypass")
_AXES = (("UP", "DOWN"), ("LEFT", "RIGHT"))


class SOCDCleaner:
    def __init__(self, mode="neutral"):
        assert mode in MODES, f"unknown SOCD mode {mode!r}"
        self.mode = mode
        self._prev = frozenset()
        self._win = {}                      # axis key -> winning direction name

    def clean(self, pressed):
        """raw pressed button names -> cleaned set (only directions are touched)."""
        pressed = set(pressed)
        out = set(pressed)
        for a, b in _AXES:
            key = a + b
            ca, cb = a in pressed, b in pressed
            pa, pb = a in self._prev, b in self._prev
            keep_a = keep_b = False

            if ca and cb:                   # SOCD conflict on this axis
                m = self.mode
                if m == "bypass":
                    keep_a = keep_b = True
                elif m == "neutral":
                    pass                    # neither
                elif m == "up_priority":
                    keep_a = (a == "UP")    # Up wins on vertical; horizontal -> neutral
                elif m == "last_win":
                    if ca and not pa and cb and not pb:
                        self._win[key] = None   # both same frame -> no winner (neutral)
                    elif ca and not pa:
                        self._win[key] = a
                    elif cb and not pb:
                        self._win[key] = b
                    keep_a = self._win.get(key) == a
                    keep_b = self._win.get(key) == b
                elif m == "first_win":
                    # first press to stand alone wins and holds; a same-frame tie from
                    # neutral has no "first" -> stays neutral until one stands alone
                    keep_a = self._win.get(key) == a
                    keep_b = self._win.get(key) == b
            elif ca:
                keep_a = True; self._win[key] = a
            elif cb:
                keep_b = True; self._win[key] = b
            else:
                self._win[key] = None

            if not keep_a:
                out.discard(a)
            if not keep_b:
                out.discard(b)

        self._prev = frozenset(pressed)
        return out


def clean_once(pressed, mode="neutral"):
    """Stateless one-shot (for neutral/up_priority/bypass)."""
    return SOCDCleaner(mode).clean(pressed)
